cmd/kubectl-accurate/sub/template_list.go

```diff
@@ -0,0 +1,98 @@
+package sub
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cybozu-go/accurate/pkg/constants"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type templateListOpts struct {
+	streams  genericclioptions.IOStreams
+	client   client.Client
+	template string
+}
+
+func newTemplateListCmd(streams genericclioptions.IOStreams, config *genericclioptions.ConfigFlags) *cobra.Command {
+	opts := &templateListOpts{}
+
+	cmd := &cobra.Command{
+		Use:     "list [TEMPLATE]",
+		Aliases: []string{"ls"},
+		Short:   "List template namespace trees hierarchically",
+		Long: `List template namespace trees hierarchically.
+If TEMPLATE is not given, all root namespaces and their children will be shown.
+If TEMPLATE is given, only the tree under the TEMPLATE namespace will be shown.`,
+		Args: cobra.MaximumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := opts.Fill(streams, config, args); err != nil {
+				return err
+			}
+			return opts.Run(cmd.Context())
+		},
+	}
+	return cmd
+}
+
+func (o *templateListOpts) Fill(streams genericclioptions.IOStreams, config *genericclioptions.ConfigFlags, args []string) error {
+	o.streams = streams
+	cl, err := makeClient(config)
+	if err != nil {
+		return err
+	}
+	o.client = cl
+
+	if len(args) > 0 {
+		o.template = args[0]
+	}
+
+	return nil
+}
+
+func (o *templateListOpts) Run(ctx context.Context) error {
+	if o.template != "" {
+		return o.showNS(ctx, o.template, 0)
+	}
+
+	templates := &corev1.NamespaceList{}
+	if err := o.client.List(ctx, templates, client.MatchingLabels{constants.LabelType: constants.NSTypeTemplate}); err != nil {
+		return fmt.Errorf("failed to list the template namespaces: %w", err)
+	}
+
+	for _, ns := range templates.Items {
+		if err := o.showNS(ctx, ns.Name, 0); err != nil {
+			return err
+		}
+	}
```

Comment:
Unlike root/sub-namespaces, the above `templates` contain both the root templates and non-root templates.
Try and see some examples.

Here, we should eliminate non-root template namespaces by referencing `accuarte.cybozu.com/template` label.

Reply:
I see, Thank you follow up.
I fixed:
https://github.com/cybozu-go/accurate/pull/9/commits/10dc7eb114611490206534e18911c6cf96446418

I checked with the following command:

<details>

```console
$ kubectl create ns temp1
$ kubectl create ns temp2
$ kubectl create ns ref1
$ kubectl create ns ref2
$ kubectl create ns ref3

$ go run ./cmd/kubectl-accurate/main.go ns set-type temp1 template
$ go run ./cmd/kubectl-accurate/main.go ns set-type ref1 template
$ go run ./cmd/kubectl-accurate/main.go ns set-type ref2 template
$ go run ./cmd/kubectl-accurate/main.go ns set-type temp2 template
$ go run ./cmd/kubectl-accurate/main.go template set ref1 temp1
$ go run ./cmd/kubectl-accurate/main.go template set ref2 ref1
$ go run ./cmd/kubectl-accurate/main.go template set ref3 temp2

$ go run ./cmd/kubectl-accurate/main.go template list
 temp1
    ⮡ref1
        ⮡ref2
 temp2
    ⮡ref3
```

</details>